contracts/Comptroller.sol

```diff
@@ -755,8 +727,29 @@ contract Comptroller is
             revert InsufficientShortfall();
         }
 
+        if ((snapshot.averageLT * (1e18 + snapshot.liquidationIncentiveAvg)) > snapshot.healthFactor) {
+            revert ToxicLiquidation();
+        }
+
+        Market storage marketCollateral = markets[vTokenCollateral];
+
+        uint256 closeFactor;
+        unchecked {
+            if (snapshot.healthFactor >= 1e18) revert InsufficientShortfall();
+            uint256 wtAvg = snapshot.averageLT;
+            if (snapshot.healthFactor >= snapshot.healthFactorThreshold) {
+                uint256 numerator = borrowBalance * 1e18 - wtAvg * snapshot.totalCollateral;
+                uint256 denominator = borrowBalance *
+                    (1e18 - ((wtAvg * (1e18 + marketCollateral.maxLiquidationIncentiveMantissa)) / 1e18));
+                closeFactor = (numerator * 1e18) / denominator;
+                closeFactor = closeFactor > 1e18 ? 1e18 : closeFactor;
+            } else {
+                closeFactor = 1e18;
+            }
```

Comment:
Maybe we could also move the calculation of the variable close factor to the LiquidationManager